toolchain/sem_ir/inst_kind.h

```diff
@@ -110,6 +110,13 @@ class InstKind : public CARBON_ENUM_BASE(InstKind) {
 #define CARBON_SEM_IR_INST_KIND(Name) CARBON_ENUM_CONSTANT_DECL(Name)
 #include "toolchain/sem_ir/inst_kind.def"
 
+  // Returns the `InstKind` for an instruction, for `CARBON_KIND_SWITCH`.
+  //
+  // The `Self` parameter is an implementation detail to allow `ForImpl` to be
+  // defined after this template, and should not be specified.
+  template <typename InstT, typename Self = InstKind>
+  static constexpr auto For = Self::template ForImpl<InstT>();
```

Comment:
Why is the indirection through a function necessary? What goes wrong if you directly use:
```c++
  template <typename InstT>
  static constexpr auto For = InstT::Kind;
```

Reply:
That tries to copy `Definition`, which doesn't work. But as I was poking at this again, I realized `auto&` might be the fix for that.